# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/tests/browser/browser_test_ContentCache.js
# Commit: b4c05874394f
# Full Hash: b4c05874394f31b10e8baca8a7a36c34809c2b82
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-02-17 21:21:27
# Description:
#   Bug 1755438 - Make `ContentCacheInParent::GetTextRect` check `mSelection` before referring it r=m_kato
#   
#   Unfortunately, the adding test cannot reproduce the crash, but I have no idea
#   how to make `mSelection` being `Nothing` until next native event loop.  Anyway,
#   it's worthwhile to add new tests.
# ==============================================================================

diff -r 554b03ae80a9 -r b4c05874394f widget/tests/browser/browser_test_ContentCache.js
--- a/widget/tests/browser/browser_test_ContentCache.js	Thu Feb 17 06:33:39 2022 +0000
+++ b/widget/tests/browser/browser_test_ContentCache.js	Thu Feb 17 08:53:35 2022 +0000
@@ -247,6 +247,56 @@
           );
         }
       })();
+
+      /**
+       * Test when no editable element has focus.
+       */
+      notifications = [];
+      await SpecialPowers.spawn(browser, [], () => {
+        content.document.body.innerHTML = "abcdef";
+      });
+
+      await waitForSendingIMENotificationsInContent();
+
+      (function() {
+        checkNotifications(
+          [
+            { type: "notify-focus", expected: false },
+            { type: "notify-blur", expected: true },
+          ],
+          "removing editor should make ContentCacheInParent not have any data"
+        );
+        const text = EventUtils.synthesizeQueryTextContent(0, 1000);
+        ok(
+          !text?.succeeded,
+          "query text content should fail because no editable element has focus"
+        );
+        const selection = EventUtils.synthesizeQuerySelectedText();
+        ok(
+          !selection?.succeeded,
+          "query selected text should fail because no editable element has focus"
+        );
+        const caret = EventUtils.synthesizeQueryCaretRect(0);
+        ok(
+          !caret?.succeeded,
+          "query caret rect should fail because no editable element has focus"
+        );
+        const textRect = EventUtils.synthesizeQueryTextRect(0, 5, false);
+        ok(
+          !textRect?.succeeded,
+          "query text rect should fail because no editable element has focus"
+        );
+        const textRectArray = EventUtils.synthesizeQueryTextRectArray(0, 5);
+        ok(
+          !textRectArray?.succeeded,
+          "query text rect array should fail because no editable element has focus"
+        );
+        const editorRect = EventUtils.synthesizeQueryEditorRect();
+        todo(
+          !editorRect?.succeeded,
+          "query editor rect should fail because no editable element has focus"
+        );
+      })();
     }
   );
 });
